### PR TITLE
Added tripcodes for persistent identity

### PR DESCRIPTION
**Pull request by nedp**
Added tripcodes to usernames to allow users to keep a persistent, secure identity between sessions.

### CODE REVIEW

**Comment by nedp** [comment]
note: force pushed an ammended typo fix - repull if needed

File: helpers/trip.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+    if (boundary === -1) {
+      return username;
+    }
+    if (boundary === username.length - 1) {
+      return username.slice(0, -1);
+    }
+
+    // If there's a second '#', treat it as the tripcode if it's the
+    // last char, otherwise use the full hash.
```
Comment (Druddigon) [inline helpers/trip.js]: This comment didn't really make sense to me. I'm interpreting it as treat that second '#' (and only that character) as the tripcode. Should it be something like the comment above (if the tripcode starts with a '#' ....)?

Comment (nedp) [inline helpers/trip.js]: Clarified the comment locally but can't push atm (git is hanging on network operations for some reason). If there's anything else I can put it all in one commit and push when my git stops throwing a fit.

File: app.js [hunk 2]
```diff
@@ -43,11 +48,16 @@ io.on('connection', function (socket) {
     // Update user list for all clients
     io.emit('update-userlist', chatUserList);
     socket.broadcast.emit('message', { message: data.username + ' has joined the chat.' });
+    socket.emit('message', {message : 'Welcome to the chat'});
```
Comment (rongduan-zhu) [inline app.js]:
```javascript
{ message: 'Welcome to the chat' });
```
No space after member, spaces for oneline objects

File: app.js [hunk 3]
```diff
@@ -43,11 +48,16 @@ io.on('connection', function (socket) {
...
+    hasJoined = true;
   });
 
-  socket.emit('message', {message : 'Welcome to the chat'});
   // When a user sends a message, send it to all other users
   socket.on('messages.send', function (data) {
+    if (!hasJoined) {
+      socket.emit('message', {message: 'connection error - please refresh the page'});
```
Comment (rongduan-zhu) [inline app.js]: same here

File: gulpfile.js [hunk 4]
```diff
@@ -11,6 +11,8 @@ var concat = require('gulp-concat'),
 
 var karma = require('karma').server;
 
+var jasmine = require('gulp-jasmine');
```
Comment (rongduan-zhu) [inline gulpfile.js]: I don't quite understand why gulp-jasmine is added? Could you please explain its purpose?

Comment (nedp) [inline gulpfile.js]: To test server side code using [jasmine via gulp](https://github.com/neroma/neroma-chat/pull/15/files#diff-b9e12334e9eafd8341a6107dd98510c9R65).

Couldn't see any tests of server code to use as a precedent (which is disturbing, btw), so I went with jasmine through gulp.

File: app.js (hunk 2 above)
Comment (nedp) [inline app.js]: awks because that was let through in a previous pull request. Maybe we should add this to the linter?

File: gulpfile.js (hunk 4 above)
Comment (rongduan-zhu) [inline gulpfile.js]: okay, isn't this outside of the scope of this PR?

**Comment by rongduan-zhu** [comment]
@nedp How do we know if the server tests are successful or not?

**Comment by nedp** [comment]
Rebased to resolve conflicts and used gulp-notify to highlight the test results.
Look for something like:
```
[Gulp notification]: Running server tests...
<details>
SUCCESS: 1 spec, 0 failures, 0 skipped, 0 disabled in 0.008s.
[Gulp notification]: Server tests done
<other tests>
```
```

**Comment by rongduan-zhu** [comment]
:+1:

**Pull request merged**